src/credentials_rotators/npm-token-rotation/src/lambda/create-new-token/index.ts

```diff
@@ -0,0 +1,202 @@
+
+  const stepFnPayload = {
+    tokenEvent: currentVersion,
+    secretArn: event.SecretId,
+  };
+  
+  await stepFnClient
+    .startExecution({
+      stateMachineArn: stepFnArn,
+      input: JSON.stringify(stepFnPayload),
+    })
+    .promise();
+
+  await secretsManagerClient
+    .updateSecretVersionStage({
+      SecretId: event.SecretId,
+      VersionStage: "AWSCURRENT",
+      MoveToVersionId: event.ClientRequestToken,
+      RemoveFromVersionId: currentVersion,
+    })
+    .promise();
+
+  console.info(
+    `end:finishSecret(${event.SecretId}, ${event.ClientRequestToken})`
+  );
+};
+
+const createSecret = async (
+  event: SecretRotationEvent,
+  credentials: NPMCredentials,
+  tokenConfig: AccessTokenItem
+) => {
+  console.info(
+    `start:createSecret(${event.SecretId}, ${event.ClientRequestToken})`
+  );
+  // check if there is already a credential that is pending
+  try {
+    await getSecret(event.SecretId, tokenConfig.secretKey, {
+      ClientRequestToken: event.ClientRequestToken,
+      stage: "AWSPENDING",
+    });
+    console.info(`Secret already is in AWSPENDING. Not creating a new one`);
+  } catch (e) {
+    if ((e as aws.AWSError).code === "ResourceNotFoundException") {
+      const secretsManagerClient = getSecretsManagerClient();
+      console.info("Requesting a new npm access token");
+
+      const newToken = await createAccessToken(
+        credentials.username,
+        credentials.password,
+        credentials.otpSeed
+      );
+      await secretsManagerClient
+        .putSecretValue({
+          SecretId: event.SecretId,
+          ClientRequestToken: event.ClientRequestToken,
+          VersionStages: ["AWSPENDING"],
+          SecretString: JSON.stringify({ [tokenConfig.secretKey]: newToken }),
+        })
+        .promise();
+      console.info("createSecret: Successfully put secret");
+    } else {
+      throw e;
+    }
+  }
+  console.info(
+    `end:createSecret(${event.SecretId}, ${event.ClientRequestToken})`
+  );
+};
+
+/**
+ * Checks the current status of the Secret and throws error if no rotation is needed
+ * @param secretArn Arn of the secret that is being rotated
+ * @param clientRequestToken clientRequestToken for the rotation
+ */
+const assertRotationStatus = async (
+  secretArn: string,
+  clientRequestToken: string
+) => {
+  console.info(
+    `start:assertRotationStatus(${secretArn}, ${clientRequestToken})`
+  );
+  const secretsMetadata = await getSecretsManagerClient()
+    .describeSecret({ SecretId: secretArn })
+    .promise();
+  if (!secretsMetadata.RotationEnabled) {
+    console.error(`Secret ${secretArn} is not enabled for rotation`);
+    throw new Error("Secret ${secretArn} is not enabled for rotation");
+  }
+  const versions = secretsMetadata.VersionIdsToStages;
+  if (!versions || !versions[clientRequestToken]) {
+    console.error(`Secret ${secretArn} has no stage for rotation of secret.`);
```

Comment:
re-consider logging ARNs

Reply:
These logs are not accessible outside the account and secret arns should be OK in the logs as to access the values one needs to have access to the account itself. Otherwise the logs would end up being useless